File: week3/이지훈/광고삽입.java
// 이해 포기..
public class 광고삽입 {
    public String solution(String play_time, String adv_time, String[] logs) {
        int playTime = convertToSecond(play_time);
        int advTime = convertToSecond(adv_time);
        long[] arr = new long[playTime + 1];

        // 동영상 시청 시작 시간, 시청 종료 시간 증감연산
        for (String log : logs) {
            String[] times = log.split("-");
            int startTime = convertToSecond(times[0]);
            int endTime = convertToSecond(times[1]);

            arr[startTime]++;
            arr[endTime]--;
        }

        for (int i = 1; i <= playTime; i++) arr[i] += arr[i - 1]; // i초에 동영상을 시청하는 인원
        for (int i = 1; i <= playTime; i++) arr[i] += arr[i - 1]; // i초 까지 동영상을 시청한 시간

        long maxTime = arr[advTime - 1];
        long startTime = 0;
        for (int i = 0; i + advTime <= playTime; i++) {
            long time = arr[i + advTime] - arr[i];

            if (time > maxTime) {
                maxTime = time;
                startTime = i + 1;
            }
        }

        return String.format("%02d:%02d:%02d", startTime / 3600, (startTime / 60) % 60, startTime % 60);
    }

    private static int convertToSecond(String time) {
        String[] times = time.split(":");
        return Integer.parseInt(times[0]) * 3600 + Integer.parseInt(times[1]) * 60 + Integer.parseInt(times[2]);
    }
}

File: week3/한지희/광고 삽입.kt
class Advertisement {
    fun solution(play_time: String, adv_time: String, logs: Array<String>): String {
        val totalSeconds = IntArray(360_000)
        val playTimeSeconds = timeFormatToSeconds(play_time)
        val advTimeSeconds = timeFormatToSeconds(adv_time)

        logs.forEach { log ->
            val screenTime = log.split("-")
            val start = timeFormatToSeconds(screenTime[0])
            val end = timeFormatToSeconds(screenTime[1])

            for (i in start until end) {
                totalSeconds[i]++
            }
        }

        var currentTime = 0L
        for (i in 0 .. advTimeSeconds) {
            currentTime += totalSeconds[i]
        }

        var maxTimeStartSeconds = 0
        var maxTime = currentTime

        for (i in advTimeSeconds + 1 .. playTimeSeconds) {
            currentTime = currentTime + totalSeconds[i] - totalSeconds[i - advTimeSeconds]
            if (currentTime > maxTime) {
                maxTime  = currentTime
                maxTimeStartSeconds = i - advTimeSeconds + 1
            }
        }

        return convertTimeFormat(maxTimeStartSeconds)
    }

    private fun timeFormatToSeconds(time: String): Int {
        val (hour, min, sec) = time.split(":").map { it.toInt() }
        return (hour * 3600) + (min * 60) + sec
    }

    private fun convertTimeFormat(time: Int): String = String.format("%02d:%02d:%02d", time / 3600, time / 60 % 60, time % 60)
}